daemon/lib/lightning/lnd/lnd.go

```diff
@@ -0,0 +1,313 @@
+	closeConnection func()
+}
+
+type Option func(*Options)
+
+func WithLndEndpoint(endpoint string) Option {
+	return func(o *Options) {
+		o.LndEndpoint = endpoint
+	}
+}
+
+func WithMacaroonFilePath(path string) Option {
+	return func(o *Options) {
+		o.MacaroonFilePath = path
+	}
+}
+
+func WithTLSCertFilePath(path string) Option {
+	return func(o *Options) {
+		o.TLSCertFilePath = path
+	}
+}
+
+func WithNetwork(network Network) Option {
+	return func(o *Options) {
+		o.Network = network
+	}
+}
+
+type Options struct {
+	LndEndpoint      string
+	MacaroonFilePath string
+	TLSCertFilePath  string
+	Network          Network
+}
+
+type Network string
+
+var (
+	Mainnet Network = "mainnet"
+	Regtest Network = "regtest"
+	Testnet Network = "testnet"
+)
+
+// NewClient creates a lnd client from a daprlndconnectURI string.
+// This Client establishes a grpc connection with a lnd node using dapr.
+// It is using two stubs: router and ln.
+func NewClient(ctx context.Context, opts ...Option) (*Client, error) {
+	// Default options
+	options := Options{
+		LndEndpoint:      "localhost:10009",
+		MacaroonFilePath: "/root/.lnd/data/chain/bitcoin/{Network}/admin.macaroon",
+		TLSCertFilePath:  "/root/.lnd/tls.cert",
+		Network:          "mainnet",
+	}
+
+	// Apply options
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	options.MacaroonFilePath = strings.Replace(options.MacaroonFilePath, "{Network}", string(options.Network), -1)
+
+	// Get the connection parameters
+	macaroonFileBytes, err := os.ReadFile(options.MacaroonFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading macaroon file: %w", err)
+	}
+
+	creds, err := credentials.NewClientTLSFromFile(options.TLSCertFilePath, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed creating TLS credentials: %w", err)
+	}
+
+	mac := &macaroon.Macaroon{}
+	err = mac.UnmarshalBinary(macaroonFileBytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed unmarshalling macaroon: %w", err)
+	}
+
+	macCred, err := macaroons.NewMacaroonCredential(mac)
+	if err != nil {
+		return nil, fmt.Errorf("failed creating macaroon credentials: %w", err)
+	}
+
+	conn, err := grpc.NewClient(options.LndEndpoint, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(macCred))
+	if err != nil {
+		return nil, fmt.Errorf("failed connecting to LND node: %w", err)
+	}
+
+	routerClient, lndClient, stateClient, invoicesClient := createInnerLNDClients(conn)
+
+	client := &Client{
+		routerClient:   *routerClient,
+		lndClient:      *lndClient,
+		stateClient:    *stateClient,
+		invoicesClient: *invoicesClient,
+		closeConnection: func() {
+			err := conn.Close()
+			if err != nil {
+				fmt.Printf("error closing connection: %v\n", err)
+			}
+		},
+	}
+
+	return client, nil
+}
+
+// AddInvoice creates an invoice in the lnd node
+func (dc *Client) AddInvoice(ctx context.Context, invoiceRequest *lnrpc.Invoice) (*lnrpc.AddInvoiceResponse, error) {
+	return dc.lndClient.AddInvoice(ctx, invoiceRequest)
+}
```

Comment:
this response depends on lnrpc, can we return something ours?

Reply:
Do you mean like creating an object of ours wrapping that response?

Reply:
yes, with just the info we will need

Reply:
> this response depends on lnrpc, can we return something ours?

hmm, I do think using LND's own response is enough, less code :D

Reply:
tbh the response object looks pretty basic imo

Reply:
wait, why do we even have `AddInvoice` for? There's already a `GenerateInvoice` method

Reply:
This is used in the GenerateInvoice method, its not in the interface

Reply:
I can do it private, but I don't think this does a big change since we are going to use the interface

Reply:
I have removed it :)